dead_pool_reaper.go

```diff
@@ -0,0 +1,122 @@
+package work
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/garyburd/redigo/redis"
+)
+
+var DeadTime = 5 * time.Minute
+
+type deadPoolReaper struct {
+	namespace string
+
+	pool *redis.Pool
+
+	stopChan         chan struct{}
+	doneStoppingChan chan struct{}
+}
+
+func newdeadPoolReaper(namespace string, pool *redis.Pool, requeueKey string, jobNames []string) *deadPoolReaper {
+	return &deadPoolReaper{
+		namespace:        namespace,
+		pool:             pool,
+		stopChan:         make(chan struct{}),
+		doneStoppingChan: make(chan struct{}),
+	}
+}
+
+func (r *deadPoolReaper) start() {
+	go r.loop()
+}
+
+func (r *deadPoolReaper) stop() {
+	close(r.stopChan)
+	<-r.doneStoppingChan
+}
+
+func (r *deadPoolReaper) loop() {
+	for {
+		select {
+		case <-r.stopChan:
+			close(r.doneStoppingChan)
+			return
+		default:
+			if err := r.reap(); err != nil {
+				logError("dead_pool_reaper.reap", err)
+			}
+		}
+	}
+}
+
+func (r *deadPoolReaper) reap() error {
+	// Get dead pools
+	deadPoolIDs, err := r.findDeadPools()
+	if err != nil {
+		return err
+	}
+
+	conn := r.pool.Get()
+	defer conn.Close()
+
+	workerPoolsKey := redisKeyWorkerPools(r.namespace)
+
+	// Cleanup all dead pools
+	for _, deadPoolID := range deadPoolIDs {
+		intID, err := strconv.ParseInt(deadPoolID, 10, 64)
+		if err != nil {
+			return err
+		}
+
```

Comment:
below there's 3 steps: 1) requeue 2) delete heartbeat 3) remove from set.

I'd do those 3 in a basic redis lock with SET resource_name my_random_value NX PX 30000. You don't actually even need to delete it aftewards